### ISSUE TITLE
Gradle Plugin: If tasks are configured eagerly configuration from DetektExtension isn't used

### ISSUE THREAD

**Issue by JLLeitschuh** [arturbosch/detekt#1181]
A very simple example way of demonstrating this issue is to modify the `DetektTaskKotlinDslTest` to this. **ALL** of the Gradle Kotlin test will fail.

```
internal class DetektTaskKotlinDslTest : Spek({

	val buildGradle = """
		|import io.gitlab.arturbosch.detekt.detekt
		|
		|plugins {
		|   `java-library`
		|	id("io.gitlab.arturbosch.detekt")
		|}
		|
		|repositories {
		|	jcenter()
		|	mavenLocal()
		|}
		|tasks.all { } // Simply add this line
		""".trimMargin()
```

## Expected Behavior
If the `Detekt` or any other task is configured, the `DetectExtension` configuration that comes later shouldn't be ignored.

## Current Behavior
If I have a `build.gradle.kts` file like this:

```kotlin
tasks.all { }
detekt {
    // Whatever configuration I have in here will be completely ignored.
}
```

## Steps to Reproduce (for bugs)
See above.

## Context
Some Gradle plugins (externally & internally) still do goofy things like collapse the configuration of all of tasks by doing things like `tasks.all { }`.

If such plugins are added to a consumers project, all of their configuration of the `DetectExtension` is completely ignored.

## Solution

**This is going to require an API breaking change for your Gradle Plugin!!!!**

You're going to need to change all of your task inputs/outputs to use the new Gradle provider API:
https://docs.gradle.org/current/userguide/lazy_configuration.html

This will require changing all of the task inputs from things like this:
```kotlin
	@Input
	@Optional
	var filters: String? = null

	@InputFile
	@Optional
	@PathSensitive(PathSensitivity.ABSOLUTE)
	var baseline: File? = null
```

To this:
```kotlin
	@Input
	@Optional
	var filters: Property<String> = objectFactory.property<String>()

	@InputFile
	@Optional
	@PathSensitive(PathSensitivity.RELATIVE)
	var baseline: RefgularFileProperty = newInputFile()
```

I **HIGHLY** recommend that you make these changes before your `1.0.0` release. Failing to do so will require a major version bump of your gradle plugin as these will be API breaking changes!

## Your Environment
```
------------------------------------------------------------
Gradle 4.10.1
------------------------------------------------------------

Build time:   2018-09-12 11:33:27 UTC
Revision:     76c9179ea9bddc32810f9125ad97c3315c544919

Kotlin DSL:   1.0-rc-6
Kotlin:       1.2.61
Groovy:       2.4.15
Ant:          Apache Ant(TM) version 1.9.11 compiled on March 23 2018
JVM:          1.8.0_92 (Oracle Corporation 25.92-b14)
OS:           Mac OS X 10.13.6 x86_64
```
* Version used: `master` of this repo
* Link to your project: This project is a great example!

**Comment by marschwar** [comment]
Thank you for pointing this out. I will look into it.

**Comment by 3flex** [comment]
@JLLeitschuh can you try adding this Gradle flag to one of your failing builds and see if that works around the issue?
`-Dorg.gradle.internal.tasks.eager=true`

If that resolves it it's an easy fix.

**Comment by marschwar** [comment]
@JLLeitschuh  In general I really like the idea of lazily evaluated properties but in combination with `tasks.all` the order still seems to be relevant. Maybe I misunderstood the problem you were describing.

To test this I changed the debug field in the extension to a `Property<Boolean>` and in the task to `Provider<Boolean>`. The configuration below still results in `debug = false`. If I switch the order it still produces `debug = true`.
```
tasks.all {
  if (it.name == 'detekt') {
  	println "debug = " + it.debug.get()
  }
}
detekt {
    debug = true
}
```
Did I miss anything? Is this even the problem you were talking about?

**Comment by JLLeitschuh** [comment]
@marschwar I think you're internal configuration of the plugin is still wrong then. What did you change `createAndConfigureDetektTask` to be?

You need to do something like this:

```kotlin
// This is the correct one!!
private fun createAndConfigureDetektTask(project: Project, extension: DetektExtension) {
	val detektTask = project.tasks.register(DETEKT, Detekt::class.java) {
		debug.set(project.provider { extension.debug })
		// .......
```

I'm guessing you were just doing this?
```kotlin
// This is incorrect!
private fun createAndConfigureDetektTask(project: Project, extension: DetektExtension) {
	val detektTask = project.tasks.register(DETEKT, Detekt::class.java) {
		debug.set(extension.debug)
		// .......
```

The second example doesn't lazily call the getter for `debug`.

**Comment by marschwar** [comment]
Neither. Since the `Property` interface extends the `Provider` it is possible to assign the extension property directly to the task.

```kotlin
// DetektExtension.kt
var debug: Property<Boolean> = project.objects.property(Boolean::class.javaObjectType)

// Detekt.kt
@Internal
@Optional
lateinit var debug: Provider<Boolean>

// DetektPlugin.kt
val detektTask = project.tasks.register(DETEKT, Detekt::class.java) {
    debug = extension.debug
```

I also tried it the way you suggested. 

```kotlin
// DetektExtension.kt
var debug: Boolean? = null

// Detekt.kt
@Internal
@Optional
val debug: Property<Boolean> = project.objects.property(Boolean::class.javaObjectType)

// DetektPlugin.kt
val detektTask = project.tasks.register(DETEKT, Detekt::class.java) {
	debug.set(project.provider { extension.debug ?: DetektExtension.DEFAULT_DEBUG_VALUE })
```

With a few println I could show, that the setter for the detekt extension is called after the `tasks.all {}` is completely processed. To for this particular case using lazy properties does not seem to be the solution. Nonetheless I think we should migrate.

**Comment by JLLeitschuh** [comment]
@marschwar Correct, sorry that I wasn't clear about that.

**Comment by 3flex** [comment]
Fixed now #1194 merged?

**Comment by JLLeitschuh** [comment]
I believe so, although I haven't tested it.
